lib/exq/enqueuer/uniqueness.ex

```diff
@@ -0,0 +1,67 @@
+defmodule Exq.Enqueuer.Uniqueness do
+  @moduledoc """
+    This module is used to enforce job uniqueness when a job is first enqueued.
+    It does this by either receiving or generating a key and saving it in redis, (the "enqueue key").
+    Finally it saves a reference to this key referenced by the job number, (the "completion key").
+    From the middleware, we remove both keys when the job completes via the remove_locks/1 function.
+  """
+
+  import Base, only: [encode16: 1]
+  import String, only: [downcase: 1]
+  alias Exq.Redis.Connection, as: Redis
+  alias Exp.Middleware.Pipeline
+
+  def remove_key(redis, key), do: Redis.del!(redis, key)
+  def hash_string(key), do: :crypto.hash(:md5, key) |> encode16 |> downcase
+
+  def with_unique_lock(callback, redis, namespace, args \\ [], key \\ "") do
+    enqueue_key = generate_enqueue_key(namespace, key)
+
+    if enqueue_key_exists?(redis, enqueue_key) do
+      {:job_not_unique, enqueue_key}
+    else
+      set_key(redis, enqueue_key)
+      {:reply, {:ok, jid}, _pipeline} = callback.()
+      completion_key = generate_completion_key(namespace, jid)
+      set_key(redis, completion_key, enqueue_key)
+      {:ok, jid}
+    end
```

Comment:
https://redis.io/commands/setnx is the preferred way to handle this kind of scenario.

get # client a
get # client b
set # client a
set # client b

This will result in two jobs. Though it might not happen with one node as the requests are serialized by the gen server, but might happen if there are more than one nodes.